### PR TITLE
Radix client

### PR DESCRIPTION
**Pull request by balda-rdx**
This PR adds a Client for the radix network.

This client makes use of the radix gateway api: https://radix-babylon-gateway-api.redoc.ly/
It also uses the radix engine toolkit: https://github.com/radixdlt/typescript-radix-engine-toolkit/tree/main and the gateway sdk: https://www.npmjs.com/package/@radixdlt/babylon-gateway-api-sdk

### CODE REVIEW

File: packages/xchain-radix/package.json [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+  "files": [
+    "lib"
+  ],
+  "repository": {
+    "type": "git",
+    "url": "git@github.com:xchainjs/xchainjs-lib.git"
+  },
+  "scripts": {
+    "clean": "rm -rf .turbo && rm -rf lib",
+    "build": "yarn clean && rollup -c",
```
Comment (0xp3gasus) [inline packages/xchain-radix/package.json:29]: Add the "build:release" script the rest of the clients have

**Comment by 0xp3gasus** [comment]
> This PR adds a Client for the radix network.
> 
> This client makes use of the radix gateway api: https://radix-babylon-gateway-api.redoc.ly/ It also uses the radix engine toolkit: https://github.com/radixdlt/typescript-radix-engine-toolkit/tree/main and the gateway sdk: https://www.npmjs.com/package/@radixdlt/babylon-gateway-api-sdk

First of all, thanks for opening the pull request!! I left you a few comments due the latest changes we have made in the repo. On the other side. Is there any way to get testnet faucet to test transactions?

**Review by 0xp3gasus — CHANGES_REQUESTED** [review]

**Comment by balda-rdx** [comment]
> > This PR adds a Client for the radix network.
> > This client makes use of the radix gateway api: https://radix-babylon-gateway-api.redoc.ly/ It also uses the radix engine toolkit: https://github.com/radixdlt/typescript-radix-engine-toolkit/tree/main and the gateway sdk: https://www.npmjs.com/package/@radixdlt/babylon-gateway-api-sdk
> 
> First of all, thanks for opening the pull request!! I left you a few comments due the latest changes we have made in the repo. On the other side. Is there any way to get testnet faucet to test transactions?

@0xp3gasus thanks for the feedback. I'll be addressing all the comments and yes, there's a way to get testnet funds from the faucet. will add it

File: packages/xchain-radix/README.md [hunk 2]
```diff
@@ -0,0 +1,207 @@
+# `@xchainjs/xchain-client`
```
[the remark this answers is not in the document]
Comment (balda-rdx) [reply packages/xchain-radix/README.md:1]: done

File: packages/xchain-radix/package.json [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+    "@xchainjs/xchain-crypto": "^0.3.1",
+    "@xchainjs/xchain-util": "^0.13.3"
+  },
+  "dependencies": {
+    "@radixdlt/babylon-gateway-api-sdk": "^1.4.1",
+    "@radixdlt/radix-engine-toolkit": "^1.0.3",
+    "bech32": "^2.0.0",
+    "bip32": "^4.0.0",
+    "ed25519-hd-key": "^1.3.0",
+    "rollup-plugin-peer-deps-external": "^2.2.4"
```
[the remark this answers is not in the document]
Comment (balda-rdx) [reply packages/xchain-radix/package.json:53]: I think it was needed with the previous structure of the repo, but might not be needed anymore. removing it

File: packages/xchain-radix/package.json (hunk 1 above)
Comment (balda-rdx) [reply packages/xchain-radix/package.json:29]: done

File: packages/xchain-radix/src/client.ts [hunk 4]
```diff
@@ -0,0 +1,852 @@
...
+}
+
+/**
+ * Custom Radix client
+ */
+
+export default class Client extends BaseXChainClient {
+  private radixSpecificClient: RadixSpecificClient
+  private curve: Curve
+  constructor(
```
Comment (0xp3gasus) [inline packages/xchain-radix/src/client.ts:426]: To be consistent with the rest of XChainJS clients, use the following type for the constructor `XChainClientParams & { curve?: Curve }`. To simplify the creation of the client, I would add 'Ed25519' as default value to the curve parameter.

File: packages/xchain-radix/src/client.ts [hunk 5]
```diff
@@ -0,0 +1,852 @@
...
+  getAddress(): string {
+    throw new Error('getAddress is synchronous and cannot retrieve addresses directly. Use getAddressAsync instead.')
+  }
+
+  /**
+   * Get the current address asynchronously for a given account.
+   * @returns {Address} A promise resolving to the current address.
+   * A phrase is needed to create a wallet and to derive an address from it.
+   */
+  async getAddressAsync(): Promise<string> {
```
Comment (0xp3gasus) [inline packages/xchain-radix/src/client.ts:537]: Add walletIndex parameter to `getAddressAsync`

File: packages/xchain-radix/src/client.ts [hunk 6]
```diff
@@ -0,0 +1,852 @@
...
+    this.radixSpecificClient.networkId = xChainJsNetworkToRadixNetworkId(network)
+  }
+
+  /**
+   * Get an estimated fee for a test transaction that involves sending
+   * XRD from one account to another
+   *
+   * @returns {Fee} An estimated fee
+   */
+  async getFees(): Promise<Fees> {
```
Comment (0xp3gasus) [inline packages/xchain-radix/src/client.ts:463]: I could not test this function in Mainnet network. This is the error I have

```ts
    Preview for fees was not successful

      265 |
      266 |     if (previewReceipt.receipt.status !== 'Succeeded') {
    > 267 |       throw new Error('Preview for fees was not successful')
          |             ^
      268 |     }
```

File: packages/xchain-radix/src/client.ts [hunk 7]
```diff
@@ -0,0 +1,852 @@
...
...aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa',
```
Comment (0xp3gasus) [inline packages/xchain-radix/src/client.ts:475]: Why is this needed? Does the message affect the fee estimation?

File: packages/xchain-radix/src/client.ts [hunk 8]
```diff
@@ -0,0 +1,852 @@
...
+    const nonFungibleBalances = this.convertResourcesToBalances(nonFungibleResources)
+    return fungibleBalances.concat(nonFungibleBalances)
+  }
+
+  private convertResourcesToBalances(
+    resources: FungibleResourcesCollectionItem[] | NonFungibleResourcesCollectionItem[],
+  ): Balance[] {
+    const balances: Balance[] = []
+    // Iterate through nonFungibleResources
+    const decimalDiff = XRD_DECIMAL - 8
```
Comment (0xp3gasus) [inline packages/xchain-radix/src/client.ts:148]: Why is this decimalDiff needed? AssetAmount and BaseAmount can handle assets with non 8 decimals

File: packages/xchain-radix/src/client.ts [hunk 9]
```diff
@@ -0,0 +1,852 @@
...
+        const asset: Asset = {
+          chain: RadixChain,
+          symbol: item.resource_address,
+          ticker: item.resource_address,
+          synth: false,
+        }
+
+        // We need to do this because item.amount can be either string or number
+        // depending on the type of resource
+        let xrdAmount: number
```
Comment (0xp3gasus) [inline packages/xchain-radix/src/client.ts:160]: Following the previous comment. This could be simplied as

```ts
       const amount = typeof item.amount === 'string' ? parseFloat(item.amount) : item.amount

        const balance: Balance = {
          asset: asset,
          amount: assetToBase(assetAmount(amount, XRD_DECIMAL)),
        }
```

File: packages/xchain-radix/src/client.ts [hunk 10]
```diff
@@ -0,0 +1,852 @@
...
+        if (typeof item.amount === 'string') {
+          xrdAmount = parseFloat(item.amount) * 10 ** decimalDiff
+        } else {
+          xrdAmount = item.amount * 10 ** decimalDiff
+        }
+
+        const balance: Balance = {
+          asset: asset,
+          amount: baseAmount(xrdAmount, XRD_DECIMAL),
```
Comment (0xp3gasus) [inline packages/xchain-radix/src/client.ts:169]: Does all resources have the same number as decimals?

File: packages/xchain-radix/src/client.ts [hunk 11]
```diff
@@ -0,0 +1,852 @@
...
+  // #region Public Methods
+  public async currentEpoch(): Promise<number> {
+    return this.innerGatewayClient.status.getCurrent().then((status) => status.ledger_state.epoch)
+  }
+
+  public async currentStateVersion(): Promise<number> {
+    return this.innerGatewayClient.status.getCurrent().then((status) => status.ledger_state.state_version)
+  }
+
+  public async fetchBalances(address: string): Promise<Balance[]> {
```
Comment (0xp3gasus) [inline packages/xchain-radix/src/client.ts:135]: Really interesting feature the non fungible balances. As the other clients do not support it, and to be consistent with them, I would split the code creating two different functions, one for fungible balances which can be retrieve from the getBalance function and another one for the non fungible part

File: packages/xchain-radix/src/client.ts [hunk 12]
```diff
@@ -0,0 +1,852 @@
...
+    for (const txn of committedTransactions) {
+      try {
+        if (
+          txn.raw_hex !== undefined &&
+          txn.confirmed_at !== null &&
+          txn.intent_hash !== undefined &&
+          txn.confirmed_at !== undefined
+        ) {
+          const transaction: Tx = await this.convertTransactionFromHex(txn.raw_hex, txn.intent_hash, txn.confirmed_at)
+          txList.txs.push(transaction)
```
Comment (0xp3gasus) [inline packages/xchain-radix/src/client.ts:670]: `total` has to be updated too

File: packages/xchain-radix/src/client.ts [hunk 13]
```diff
@@ -0,0 +1,852 @@
...
+
+    while (hasNextPage) {
+      const response = await this.radixSpecificClient.gatewayClient.stream.innerClient.streamTransactions({
+        streamTransactionsRequest: {
+          affected_global_entities_filter: [params.address],
+          limit_per_page: params.limit && params.limit > 100 ? 100 : params.limit,
+          from_ledger_state: {
+            state_version: params.offset,
+          },
+          manifest_resources_filter: [params.asset || ''],
```
Comment (0xp3gasus) [inline packages/xchain-radix/src/client.ts:647]: `manifest_resources_filter: params.asset ? [params.asset] : undefined,`

File: packages/xchain-radix/src/client.ts [hunk 14]
```diff
@@ -0,0 +1,852 @@
...
+    )
+    return filteredBalances
+  }
+
+  /**
+   * Get transaction history of a given address with pagination options.
+   * @param {TxHistoryParams} params The options to get transaction history. (optional)
+   * @returns {TxsPage} The transaction history.
+   */
+  async getTransactions(params: TxHistoryParams): Promise<TxsPage> {
```
Comment (0xp3gasus) [inline packages/xchain-radix/src/client.ts:633]: Is `offset` parameter mandatory?

File: packages/xchain-radix/package.json [hunk 15]
```diff
@@ -0,0 +1,54 @@
...
+  "scripts": {
+    "clean": "rm -rf .turbo && rm -rf lib",
+    "build": "yarn clean && rollup -c",
+    "build:release": "yarn exec rm -rf release && yarn pack && yarn exec \"mkdir release && tar zxvf package.tgz --directory release && rm package.tgz\"",
+    "test": "jest",
+    "compile": "tsc -p tsconfig.build.json",
+    "lint": "eslint \"{src,__tests__}/**/*.ts\" --fix --max-warnings 0",
+    "fund": "ts-node src/utils.ts"
+  },
+  "devDependencies": {
```
Comment (0xp3gasus) [inline packages/xchain-radix/package.json:36]: Remove devDependencies

**Review by 0xp3gasus — CHANGES_REQUESTED** [review]
I could not test transfers in Testnet network due to an error on the getFees

File: packages/xchain-radix/src/client.ts (hunk 7 above)
Comment (balda-rdx) [reply packages/xchain-radix/src/client.ts:475]: yes the message length does affect the fee estimation

File: packages/xchain-radix/src/client.ts (hunk 8 above)
Comment (balda-rdx) [reply packages/xchain-radix/src/client.ts:148]: got it

File: packages/xchain-radix/src/client.ts (hunk 9 above)
Comment (balda-rdx) [reply packages/xchain-radix/src/client.ts:160]: makes sense to me

File: packages/xchain-radix/src/client.ts (hunk 10 above)
Comment (balda-rdx) [reply packages/xchain-radix/src/client.ts:169]: not necessarily. fixed

**Comment by balda-rdx** [comment]
> I could not test transfers in Testnet network due to an error on the getFees

there's an e2e test that transfer funds in the testnet network, maybe take a look at that one? https://github.com/xchainjs/xchainjs-lib/pull/1146/files#diff-86a40cb25a0073784030ae41e4080be5316ec62d4291712315fd94916e2d1582R404

File: packages/xchain-radix/src/client.ts (hunk 4 above)
Comment (balda-rdx) [reply packages/xchain-radix/src/client.ts:426]: done

File: packages/xchain-radix/src/client.ts (hunk 14 above)
Comment (balda-rdx) [reply packages/xchain-radix/src/client.ts:630]: I made it optional now

File: packages/xchain-radix/src/client.ts (hunk 11 above)
Comment (balda-rdx) [reply packages/xchain-radix/src/client.ts:135]: @0xp3gasus I can totally split it but what would you like the getBalance function to return? just fungible tokens balances? that would be sort of inaccurate as non fungible tokens do contribute to the account balance

Comment (0xp3gasus) [reply packages/xchain-radix/src/client.ts:135]: EVM chains has also NFTs, but only ERC20 balances are returned. I suggest this change to be consistent with them and create a specific feature for NFTs for those clients which are compatible with them, for example, `getNFTBalances`

File: packages/xchain-radix/src/client.ts [hunk 16]
```diff
@@ -0,0 +1,853 @@
...
+          synth: false,
+        }
+
+        // We need to do this because item.amount can be either string or number
+        // depending on the type of resource
+        const amount = typeof item.amount === 'string' ? parseFloat(item.amount) : item.amount
+
+        const balance: Balance = {
+          asset: asset,
+          amount: assetToBase(assetAmount(amount)),
```
Comment (0xp3gasus) [inline packages/xchain-radix/src/client.ts:164]: ```assetToBase(assetAmount(amount, XRD_DECIMAL))```

File: packages/xchain-radix/src/client.ts (hunk 10 above)
Comment (0xp3gasus) [reply packages/xchain-radix/src/client.ts]: Is there a way to retrieve the decimals? For non 8 decimals, the number of decimals should be set, otherwise amount will not be correct

File: packages/xchain-radix/src/client.ts [hunk 17]
```diff
@@ -157,16 +157,11 @@ export class RadixSpecificClient {
...
-          xrdAmount = parseFloat(item.amount) * 10 ** decimalDiff
-        } else {
-          xrdAmount = item.amount * 10 ** decimalDiff
-        }
+        const amount = typeof item.amount === 'string' ? parseFloat(item.amount) : item.amount
 
         const balance: Balance = {
           asset: asset,
-          amount: baseAmount(xrdAmount, XRD_DECIMAL),
+          amount: assetToBase(assetAmount(amount)),
```
Comment (0xp3gasus) [inline packages/xchain-radix/src/client.ts:164]: Is there an easy way to retrieve the decimals from the assets?

File: packages/xchain-radix/src/client.ts (hunk 11 above)
Comment (balda-rdx) [reply packages/xchain-radix/src/client.ts:135]: fair enough. I'm splitting this

File: packages/xchain-radix/src/client.ts (hunk 17 above)
Comment (balda-rdx) [reply packages/xchain-radix/src/client.ts]: this should be now fixed

**Review by 0xp3gasus — APPROVED** [review]
Great work!! 🚀🚀

**Review by 0xp3gasus — APPROVED** [review]
